src/rideables/MontageStack.hpp

```diff
@@ -74,16 +112,22 @@ optional<T> MontageStack<T>::pop(int tid)
     StackNode *new_node;
     StackNode *old_node;
     optional<T> res = {};
-    do
+    while(true)
     {
         old_node = top.load();
         if (old_node == NULL)
         {
             tracker.end_op(tid);
             return res;
         }
+        begin_op();
         new_node = old_node->next;
-    } while (!top.compare_exchange_weak(old_node, new_node));
+        if (!top.CAS_verify(this, old_node, new_node))){
+            end_op();
```

Comment:
`payload` in `old_node` needs to be `pretire`-d, since we want pop happening in the epoch fetched by `begin_op` really happens persistently in this epoch.